Introduce separate Container concept for computer outputs

ctrueden opened the pull request:
Computer outputs aren't exactly mutable in the same way as inplace arguments are. Rather, they are preallocated empty containers whose contents will be filled in by the invoked computation.

My strong intuition is that having this distinction baked into the type annotations, and accessible at runtime, may be somehow helpful later.
    
This also adds a corresponding `ItemIO.CONTAINER` value for struct members; the `ItemIO.BOTH` value is retired in favor of `ItemIO.MUTABLE`, which corresponds to the `@Mutable` annotation for the `Inplace` interfaces.

ctrueden committed:
Remove redundant @Mutable annotations

It's already specified in the implemented interface,
and therefore implicit for the implementing classes.

ctrueden committed:
Introduce separate @Container for computer outputs

They aren't exactly mutable in the same was as the mutable inplace
arguments are. Rather, they are preallocated empty containers whose
contents will be filled in by the invoked computation.

My strong intuition is that having this distinction baked into the type
annotations, and accessible at runtime, may be somehow helpful later.

This also adds a corresponding ItemIO.CONTAINER value for struct
members; the ItemIO.BOTH value is retired in favor of ItemIO.MUTABLE,
which corresponds to the @Mutable annotation for the Inplace interfaces.

ctrueden committed:
Mutable: improve javadoc

Explain what mutable means, and link to the Inplaces interfaces,
which are the ones making use of this annotation.

ctrueden committed:
Remove superfluous ItemIO assignments

The default value of ItemIO.AUTO should be able to figure these out.

imagejan commented on scijava/scijava-ops/src/main/java/org/scijava/param/ParameterStructs.java:
```diff
@@ -325,14 +324,16 @@ public static Method singularAbstractMethod(Class<?> functionalInterface) {
+	 * Returns a list of {@link FunctionalMethodType}s describing the input and
+	 * output types of the functional method of the specified functional type. In
+	 * doing so, the return type of the method will me marked as
+	 * {@link ItemIO#OUTPUT} and the all method parameters as {@link ItemIO#OUTPUT},
```
```suggestion
	 * {@link ItemIO#OUTPUT} and the all method parameters as {@link ItemIO#INPUT},
```

hinerm commented on scijava/scijava-ops/src/main/java/org/scijava/ops/function/Computers.java:
```diff
@@ -431,7 +431,7 @@ default void accept(final I1 in1, final I2 in2, final I3 in3, final I4 in4, fina
 	{
 
-		void compute(I1 in1, I2 in2, I3 in3, I4 in4, I5 in5, I6 in6, I7 in7, I8 in8, I9 in9, I10 in10, I11 in11, I12 in12, I13 in13, I14 in14, I15 in15, I16 in16, @Mutable O out);
+		void compute(I1 in1, I2 in2, I3 in3, I4 in4, I5 in5, I6 in6, I7 in7, I8 in8, I9 in9, I10 in10, I11 in11, I12 in12, I13 in13, I14 in14, I15 in15, I16 in16, @Container O out);
```
Overall I think `@Mutable` and `@Container` are more descriptive, but I don't understand why some compute methods have an `@Container` but others don't?

e.g. https://github.com/scijava/incubator/pull/27/files#diff-199835d92f5a48202497fb4a82748776e19fe8d432e17c609dc462708ce36c11R558

gselzer replied:
@hinerm if I remember correctly, we only need these annotations on our functional interface methods (since this is where our API will pull them from). Your link goes to a method that is overriding the interface's method, thus it is probably best to leave the annotation off.

Do you have a rule for annotation application that you would prefer?

hinerm replied:
@gselzer No that makes sense. I just noticed that the implementation methods had previously been annotated `@Mutable` so the drop in annotation stood out.

gselzer replied:
Gotcha, thanks for pointing it out, but also that it is a good change (I think you agree with me on this, more just want to record my thoughts). I would personally want these annotations required (even allowable) in as few places as possible because it is:
* annoying to have to remember to put them there
* more work to change if/when we decide to alter the annotations/add new functional interfaces and support for them

gselzer committed:
Throw IAE when no discoverable functional method

Null returns can be misleading, and could give the impression that this
is a side-effects-based Op (such as a Consumers.Arity0).

gselzer commented:
@ctrueden as far as I can see, looks good to me!

I did see a minor issue with the `null` return on `ParameterStructs.findFunctionalMethodTypes`, so I added a commit to throw an `IAE` instead.

ctrueden commented:
Shall we merge it?

ctrueden commented:
I merged it. :smile:
